File: Python/CodingPractice/LeetCode/SpiralMatrix.py
class SpiralMatrix(object):

    def main(self):
        print(self.spiralOrder([[1,2,3], [4, 5, 6],[7, 8, 9]]))
        

    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if matrix is None or len(matrix) == 0:
            return []
        
        result = []
        left = 0
        right = len(matrix[0]) - 1
        top = 0
        bottom = len(matrix) - 1
        
        while True:
            for i in range(left, right + 1):
                result.append(matrix[top][i])
            
            top +=1
            if self.isOverlap(top, bottom, left, right):
                break
                
            for i in range(top, bottom + 1):
                result.append(matrix[i][right])
                
            right -=1
            
            if self.isOverlap(top, bottom, left, right):
                break
            
            for i in range(right, left - 1, -1):
                result.append(matrix[bottom][i])
                
            bottom -=1
            
            if self.isOverlap(top, bottom, left, right):
                break
                
            for i in range (bottom, top - 1, -1):
                result.append(matrix[i][left])
            
            left +=1
            
            if self.isOverlap(top, bottom, left, right):
                break
                
        return result                            
                
    def isOverlap(self, top, bottom, left, right):
        if left > right or bottom < top:
            return True
        
        return False
        
        
        

File: Java/CodingPractice/src/main/java/Arrays/Lists/spiralMatrix.java
package Arrays.Lists;

import java.util.ArrayList;
import java.util.List;

public class spiralMatrix {


    public static void main(String arg[]){

        int[][] matrix = {{1,2,3},{4,5,6},{7,8,9}};

        List<Integer> result = spiralForm(matrix);

        for(int i : result){

            System.out.print(i);
        }

    }

    public static List<Integer> spiralForm(int[][] matrix){

        List<Integer> result = new ArrayList();

        int top = 0;
        int bottom = matrix.length - 1;
        int left = 0;
        int right = matrix[0].length - 1;


        while(true){

            for(int i = left ; i <=right ; i++){
                result.add(matrix[top][i]);

            }
            top++;

            if(isCrossed(left,right,top , bottom)){

                break;
            }

            for(int j = top ; j <= bottom ; j++ ){

                result.add(matrix[j][right]);

            }
            right--;

            if(isCrossed(left,right,top , bottom)){

                break;
            }

            for(int i = right ; i >= left ; i--){

                result.add(matrix[bottom][i]);
            }
            bottom--;
            if(isCrossed(left,right,top , bottom)){

                break;
            }

            for(int j = bottom ; j >= top ; j--){

                result.add(matrix[j][left]);
            }
            left++;
            
            if(isCrossed(left,right,top , bottom)){

                break;
            }


        }
        return result;

    }

    public static  boolean isCrossed(int left, int right, int top, int bottom){

        if(left > right || bottom < top){
            return true;
        }
        else{
            return false;
        }
    }
}
